### PR TITLE
[Merged by Bors] - feat: bijective_iff_map_univ_eq_univ

### PR DESCRIPTION
**Pull request by BoltonBailey**
For functions on finite sets, they are bijections iff they map universes into universes.

Co-authored-by: Junyan Xu <junyanxu.math@gmail.com>
Co-authored-by: Sebastian Zimmer<sebastian.zim@googlemail.com>

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by sebzim4500** [comment]
I saw this was marked 'please-adopt', so I made a new PR which fills out the implementation [here](https://github.com/leanprover-community/mathlib4/pull/7365)

**Comment by BoltonBailey** [comment]
> I saw this was marked 'please-adopt', so I made a new PR which fills out the implementation [here](https://github.com/leanprover-community/mathlib4/pull/7365)

Sweet! Thank you. Do you have permission to push to non-master branches? If not you can request that on the [zulip](https://leanprover.zulipchat.com/#narrow/stream/113489-new-members/topic/GitHub.20permission)

File: Mathlib/Data/Fintype/Basic.lean [hunk 1]
```diff
@@ -1229,6 +1229,48 @@ theorem map_univ_val_equiv (e : α ≃ β) :
     map e univ.val = univ.val := by
   rw [←congr_arg Finset.val (Finset.map_univ_equiv e), Finset.map_val, Equiv.coe_toEmbedding]
 
+/-- For functions on finite sets, they are bijections iff they map universes into universes. -/
+@[simp]
+theorem bijective_iff_map_univ_eq_univ (f : α → β) :
+    Function.Bijective f ↔ map f (Finset.univ : Finset α).val = univ.val := by
+  constructor
+  {
```
Comment (BoltonBailey) [inline Mathlib/Data/Fintype/Basic.lean:1237]: In Lean 4, we don't use curly braces to delimit code blocks anymore, instead as the style guide says

> When new goals arise as side conditions or steps, they are indented and preceded by a focusing dot · (inserted as \.); the dot is not indented.

**Comment by eric-wieser** [comment]
maintainer merge

I pushed a golf, so someone else should probably merge

File: Mathlib/Data/Fintype/Basic.lean [hunk 2]
```diff
@@ -1217,7 +1217,7 @@ end Trunc
 
 namespace Multiset
 
-variable [Fintype α] [DecidableEq α] [Fintype β]
+variable [Fintype α] [DecidableEq α] [Fintype β] [DecidableEq β]
```
Comment (j-loreaux) [inline Mathlib/Data/Fintype/Basic.lean:1220]: Either this changed `count_univ` and `map_univ_val_equiv`, or else Lean is automatically purging this hypothesis from those declarations. Either way, this (these?) `DecidableEq` hypotheses shouldn't be included on this line.

File: Mathlib/Data/Fintype/Basic.lean [hunk 3]
```diff
@@ -1229,6 +1229,32 @@ theorem map_univ_val_equiv (e : α ≃ β) :
...
+      simp [←hc]
+    have hb : b = c
+    · rw [←Multiset.mem_singleton]
+      simp [←hc, fab]
+    rw [ha, hb]
+  · intro b
+    refine (h b).imp fun a ha => ?_
+    have ha2 : a ∈ Multiset.filter (b = f ·) univ.val
+    · rw [ha, Multiset.mem_singleton]
+    simpa only [Multiset.mem_filter, mem_val, mem_univ, true_and, eq_comm] using ha2
```
Comment (alreadydone) [inline Mathlib/Data/Fintype/Basic.lean:1256]: golf
```suggestion
  exact fun eq ↦ ⟨fun _ _ ↦ inj_on_of_nodup_map (eq.symm ▸ univ.2) _ (mem_univ _) _ (mem_univ _),
    fun b ↦ have ⟨a, _, h⟩ := mem_map.mp (eq.symm ▸ mem_univ_val b); ⟨a, h⟩⟩
```

Comment (alreadydone) [inline Mathlib/Data/Fintype/Basic.lean:1256]: Updated golf, thanks to `map_univ_equiv` that Moogle found for me :)
```suggestion
    Function.Bijective f ↔ map f (Finset.univ : Finset α).val = univ.val :=
  ⟨fun bij ↦ congr_arg (·.val) (map_univ_equiv <| Equiv.ofBijective f bij),
    fun eq ↦ ⟨fun a₁ a₂ ↦ inj_on_of_nodup_map (eq.symm ▸ univ.2) a₁ (mem_univ a₁) a₂ (mem_univ a₂),
    fun b ↦ have ⟨a, _, h⟩ := mem_map.mp (eq.symm ▸ mem_univ_val b); ⟨a, h⟩⟩⟩
```

File: Mathlib/Data/Fintype/Basic.lean (hunk 2 above)
Comment (eric-wieser) [reply Mathlib/Data/Fintype/Basic.lean]: There was thankfully no incorrectly-typed lemmas here, but this was indeed misleading; I've tidied it up.

**Review by eric-wieser — APPROVED** [review]
bors merge

@alreadydone completely golfed away my changes, so I don't think I need to worry about merging my own code anyway more. Thanks!

**Pull request closed without merging**